### PR TITLE
IDTReeS - Clip boxes outside of image bounds

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR
- clips the bounding box coordinates to the image size (essentially clips boxes that go outside of the image bounds) and filters any boxes that have zero area after clipping.
- fixed bug where id of some geometries were being overriden resulting in some images have no objects.

Fixes part of #684

For `MLBS_5` the boxes before

```python
tensor([[107, -57, 165, -10],
        [157, -48, 190, -10],
        [127,  24, 198,  92],
        [149, 122, 199, 174],
        [ 30,  70,  70, 109],
        [ -6, -80,  48, -10],
        [ 47, 121, 114, 197],
        [119,  84, 144, 129],
        [108,  74, 129, 101],
        [ 36,  21,  83,  73],
        [137,  73, 187, 128],
        [ 38, -63,  81, -10],
        [ 19,  93,  62, 152],
        [ 71,  81, 117, 133]])
```

and after filtering

```python
tensor([[127,  24, 198,  92],
        [149, 122, 199, 174],
        [ 30,  70,  70, 109],
        [ 47, 121, 114, 197],
        [119,  84, 144, 129],
        [108,  74, 129, 101],
        [ 36,  21,  83,  73],
        [137,  73, 187, 128],
        [ 19,  93,  62, 152],
        [ 71,  81, 117, 133]])
```

Visual result:
image

### CODE REVIEW

**Comment by ashnair1** [comment]
The coordinate swap was already done in #683

**Comment by isaaccorley** [comment]
@ashnair1 Thanks for pointing that out. Totally missed it.

File: torchgeo/datasets/idtrees.py [hunk 1]
```diff
@@ -296,6 +296,7 @@ def _load_boxes(self, path: str) -> Tensor:
                 boxes.append([xmin, ymin, xmax, ymax])
 
         tensor = torch.tensor(boxes)
+        tensor = torch.clamp(tensor, min=0, max=self.image_size[0])
```
Comment (weiji14) [inline torchgeo/datasets/idtrees.py:299]: This assumes that images are always square (i.e. `self.image_size[0] == self.image_size[1]`). Is this always the case for this IDTReeS dataset?

Thinking about whether [`torchvision.ops.clip_boxes_to_image`](https://pytorch.org/vision/0.13/generated/torchvision.ops.clip_boxes_to_image.html#torchvision.ops.clip_boxes_to_image) would work here, or if it's better to keep things simple.

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:299]: Yes, if you check load_image we read them in as 200x200.

I'm indifferent on using clip_boxes_to_image since it appears to perform the same functionality. I would just make sure it's not a recent addition to torchvision so it makes our dependencies strict.

Comment (weiji14) [reply torchgeo/datasets/idtrees.py:299]: Cool, if the images are always square, then this addition is ok. Just hope the next person adding an object detection dataset is careful when copying this code :laughing:

**Review by weiji14 — APPROVED** [review]

File: torchgeo/datasets/idtrees.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:299]: I also had the same question about square vs rectangular, can you add a comment saying the image will always be square?

**Comment by isaaccorley** [comment]
Created generic `torchgeo.datasets.utils.filter_boxes` function in case this occurs in other object detection datasets. Also refactored to not hardcode the image size but to use the current sample image (h, w).

File: torchgeo/datasets/idtrees.py [hunk 2]
```diff
@@ -211,10 +211,20 @@ def __getitem__(self, index: int) -> Dict[str, Tensor]:
         if self.split == "test":
             if self.task == "task2":
                 sample["boxes"] = self._load_boxes(path)
+                w, h = sample["image"].shape[1:]
+                sample["boxes"], _ = filter_boxes(
+                    image_size=(h, w), boxes=sample["boxes"]
```
Comment (weiji14) [inline torchgeo/datasets/idtrees.py:216]: Doesn't matter if it's square, but should it both be `h, w`?

```suggestion
                h, w = sample["image"].shape[1:]
                sample["boxes"], _ = filter_boxes(
                    image_size=(h, w), boxes=sample["boxes"]
```

File: torchgeo/datasets/idtrees.py [hunk 3]
```diff
@@ -211,10 +211,20 @@ def __getitem__(self, index: int) -> Dict[str, Tensor]:
...
+                )
         else:
             sample["boxes"] = self._load_boxes(path)
             sample["label"] = self._load_target(path)
 
+            w, h = sample["image"].shape[1:]
+            sample["boxes"], sample["label"] = filter_boxes(  # type:ignore[assignment]
+                image_size=(h, w), boxes=sample["boxes"], labels=sample["label"]
```
Comment (weiji14) [inline torchgeo/datasets/idtrees.py:224]:
```suggestion
            h, w = sample["image"].shape[1:]
            sample["boxes"], sample["label"] = filter_boxes(  # type:ignore[assignment]
                image_size=(h, w), boxes=sample["boxes"], labels=sample["label"]
```

File: torchgeo/datasets/utils.py [hunk 4]
```diff
@@ -707,3 +709,32 @@ def percentile_normalization(
...
+    Args:
+        image_size: tuple of (height, width) of image
+        min_size: filter boxes that have any side less than min_size
+        boxes: [N, 4] shape tensor of xyxy bounding box coordinates
+        labels: (Optional) [N,] shape tensor of bounding box labels
+
+    Returns:
+        a tuple of filtered boxes and labels
+
+    .. versionadded:: 0.3.1
```
[the remark this answers is not in the document]
Comment (isaaccorley) [reply torchgeo/datasets/utils.py:731]: I could put it directly in the dataset as a method with a note to move to it's own function in 0.4.0. Does that work?

[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchgeo/datasets/utils.py:731]: That's fine by me. I wouldn't move it to `utils` unless multiple datasets need it.

[the remark this answers is not in the document]
Comment (isaaccorley) [reply torchgeo/datasets/utils.py:731]: Okay sounds good

File: torchgeo/datasets/idtrees.py (hunk 2 above)
Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py]: Fixed

File: torchgeo/datasets/idtrees.py (hunk 3 above)
Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py]: Fixed

File: torchgeo/datasets/idtrees.py [hunk 5]
```diff
@@ -391,6 +404,35 @@ def _load_geometries(self, directory: str) -> Dict[int, Dict[str, Any]]:
...
+        Args:
+            image_size: tuple of (height, width) of image
+            min_size: filter boxes that have any side less than min_size
+            boxes: [N, 4] shape tensor of xyxy bounding box coordinates
+            labels: (Optional) [N,] shape tensor of bounding box labels
+
+        Returns:
+            a tuple of filtered boxes and labels
+
+        .. versionadded:: 0.3.1
```
Comment (adamjstewart) [inline torchgeo/datasets/idtrees.py:425]: You can remove this line

File: torchgeo/datasets/idtrees.py [hunk 6]
```diff
@@ -383,14 +396,44 @@ def _load_geometries(self, directory: str) -> Dict[int, Dict[str, Any]]:
...
         for path in filepaths:
             with fiona.open(path) as src:
-                for i, feature in enumerate(src):
+                for feature in src:
                     if self.split == "train":
                         features[feature["properties"]["id"]] = feature
                     # Test set task 2 has no id
                     else:
-                        features[i] = feature
+                        id = len(features) + 1
```
Comment (isaaccorley) [inline torchgeo/datasets/idtrees.py:404]: Found a phantom bug here making some of the boxes be empty due to being overriden and not existing in `self.geometries`

File: torchgeo/datasets/idtrees.py [hunk 7]
```diff
@@ -383,14 +396,42 @@ def _load_geometries(self, directory: str) -> Dict[int, Dict[str, Any]]:
...
+                        features[id] = feature
         return features
 
+    def _filter_boxes(
```
Comment (adamjstewart) [inline torchgeo/datasets/idtrees.py:408]: I think the correct way to do this is the unfortunately verbose:
```python
from typing import override

@override
def _filter_boxes(self, image_size: Tuple[int, int], boxes: Tensor, min_size: int, labels: Tensor) -> Tuple[Tensor, Tensor]:
    ...

@override 
def _filter_boxes(self, image_size: Tuple[int, int], boxes: Tensor, min_size: int, labels: None) -> Tuple[Tensor, None]:
    ...
```
followed by the same code below. Basically, if the input labels is Tensor, the output labels is also Tensor, and if it's None it's None. Then you can remove the `type: ignore` above.

File: torchgeo/datasets/idtrees.py (hunk 6 above)
Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:404]: I'm gonna be honest, I don't really understand what `_load_geometries` does and specifically what `features` is supposed to contain. It seems like a nested dict, but what are the keys and what are the values? Also, I see that we're using `Any` and `cast` in several places, which is usually bad. Can you elaborate on this and maybe add more comments documenting it? I want to make sure you're not the only one who understands how this Dataset works.

**Review by weiji14 — DISMISSED** [review]
Code logic and docstrings seems to be sound. Did you check the image and box alignment? I'm not quite sure how to interpret the sample image you posted at https://github.com/microsoft/torchgeo/pull/760#issue-1362426866, the colors look a bit weird :sweat_smile:

**Comment by isaaccorley** [comment]
> #684

Yes. I don't select the colors. torchvision's `draw_bounding_boxes` will randomly select box colors based on the number of classes in the image since I don't provide any colors.

File: torchgeo/datasets/idtrees.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:404]: It's a dictionary of geojson like features (basically a dict with properties and a geometry). For the train set these geometries have unique ids inside of the shapely file. However, for the test set they have no ids, so I have to make one up. In this case I just make one that increments every time a new feature is added.

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: Added more comments for clarity

**Comment by isaaccorley** [comment]
Just wanted to add some notes for everyone's clarity on the dataset.

- The test set - task 1 has no boxes or labels only RGB/HSI/CHM images and LAS point clouds
- The test set - task 2 shapefiles (e.g. task2/ITC/test_MLBS.shp) and contains the image file the boxes belong to in the properties, e.g. below in `properties["plotID"]` the image file is `MLBS_1.tif`.

```python
{
  'geometry': {'coordinates': [[(541886.6194198467, 4136580.9452913683),
                                 (541886.6194198467, 4136588.0890556537),
                                 (541891.3025542134, 4136588.0890556537),
                                 (541891.3025542134, 4136580.9452913683),
                                 (541886.6194198467, 4136580.9452913683)]],
                'type': 'Polygon'},
   'id': '0',
   'properties': OrderedDict([('indvdID', 'MLBSE00048'),
                              ('plotID', 'MLBS_1.tif')]),
   'type': 'Feature'
}
```

- The train set shapefiles (e.g. train/ITC/train_MLBS.shp) contain the boxes but **do not** contain the image file. However it does contain a unique geometry id `properties["id"]`.

```python
{
  'geometry': {'coordinates': [[(542071.1868343, 4134996.0734410095),
                                 (542071.1868343, 4134999.0),
                                 (542074.96994003, 4134999.0),
                                 (542074.96994003, 4134996.0734410095),
                                 (542071.1868343, 4134996.0734410095)]],
                'type': 'Polygon'},
   'id': '0',
   'properties': OrderedDict([('id', 57), ('indvdID', 'MLBS01681')]),
   'type': 'Feature'
}
```

The image <-> geometry id mapping is inside `train/Field/itc_rsFile.csv`.


```python
"id","indvdID","rsFile"
57,"MLBS01681","MLBS_1.tif"
45,"MLBS01679","MLBS_1.tif"
49,"MLBS01674","MLBS_1.tif"
63,"MLBS01678","MLBS_1.tif"
...
```

Note that the test set **does not** have any csv files only the train set. Because of this, the implementation for parsing the dataset can become quite verbose and requires different logic for the both train set/test set - task1/ test set - task2. 

Hopefully this clarifies some of the implementation decisions.

File: torchgeo/datasets/idtrees.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: - `self.images` is a globbed list of image paths
- `self.geometries` is a dict where each entry maps a unique id to a geometry and it's properties. (test set - task 1 does not have any geometries so it's `None` for this case.
- `self.labels` is the `itc_rsFile.csv` file read into a pandas dataframe (this is only valid for the train set, so for all other cases it is `None`)

That's correct, since there are multiple shapefiles, the test set geometries were being overriden.

My initial plan was to use the `properties["indvdID"]` however I found that there were duplicates so I couldn't use this as the id so the next option was to just make a dummy id.

We could use further optimize by using a counter `i = 0` but `self._load_geometries` is only called once on instantiation of the dataset so the loop only occurs once. If you prefer I add this then that is fine with me.

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:409]: Do we need a dict for geometries? Could we use a list instead?

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: The dict makes it easy for me to grab the ids and then just index the dict for the geometries needed for the image. I could revamp the dataset and just have a list of [image_path, list_of_geometries_per_image, and list_of_labels_per_image] instead?

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:409]: I think I still don't understand. If `self.images`, `self.geometries`, and `self.labels` don't share the same keys/order, how do you map from one to the other? I'm trying to follow the path from `__getitem__` index to file loading but there are a lot of moving parts.

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: getitem[idx] -> image = self.images[idx]  -> self.labels[image] -> ids -> self.labels AND self.geometries

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: I agree this likely needs a refactor to simplify things however, this PR is for bug fixes. The dataset currently works fine as is. Would you like to refactor for 0.4.0?

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:409]: Okay, but if `self.geometries[image]`, then why does it feel like the id in `features[id]` is completely arbitrary and not the name of the image file?

Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: The id is only arbitrary for the test set not the train set. You need the id to index the csv file to find the label and image filename.

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:409]: Wait, so `self._geometries()` is only called on the test set? Because it looks like the id is arbitrary any time this function is called...

Comment (adamjstewart) [reply torchgeo/datasets/idtrees.py:409]: Nvm, I just noticed the if statement.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Okay, I think I finally understand the code well enough to review this now!

So if I understand correctly, this PR only solves issue (3) in #684. Once this PR is merged, A) all bounding boxes will be clipped to the image bounds which will fix plotting, and B) bounding boxes won't be accidentally deleted from the testing set due to integer overlap.

My only change request is to replace the current `id = len(features) + 1` logic with a counter `i = 0, i += 1` so we aren't computing the length of a list in each iteration.

File: torchgeo/datasets/idtrees.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/datasets/idtrees.py:409]: I can follow up with a refactor in another PR. I think what I will do is on instantiation do all the ugly parsing and create a list of tuples of (image_path, boxes, box_labels) and that should be less convoluted when getitem is actually called.

**Comment by isaaccorley** [comment]
> Okay, I think I finally understand the code well enough to review this now!
> 
> So if I understand correctly, this PR only solves issue (3) in #684. Once this PR is merged, A) all bounding boxes will be clipped to the image bounds which will fix plotting, and B) bounding boxes won't be accidentally deleted from the testing set due to integer overlap.
> 
> My only change request is to replace the current `id = len(features) + 1` logic with a counter `i = 0, i += 1` so we aren't computing the length of a list in each iteration.

I've made the requested change. Also note that this PR doesn't just fix plotting but also allows a user to actually train on the dataset whereas before you would have issues dealing with boxes with negative values.

**Review by adamjstewart — APPROVED** [review]
